### PR TITLE
Separated out ConnectedComponents and Coreness traits

### PR DESCRIPTION
**Pull request by bogdanstate**
`SimpleUndirectedGraph` has been growing enormously, and a refactor seems in order. It feels logical to divide different problem areas around graph mining using trait composition. This has the additional benefit of allowing reuse of the same algorithms for multiple kinds of graphs.

An important part of this change required the introduction of new base traits for `Node` and `NodeEdge`, the introduction of a `NodeType` associate type to `GraphBase`, with `TypedGraph` constraining the implementation of `NodeType = Node`, to avoid an even more involved refactor.

The general pattern I would like to move towards is having new algorithms implemented in separate traits, that nonetheless attach to `SimpleUndirectedGraphs`. This will also help us draw a neat dependency graph (heh) between the various traits, where for instance we can see that `Coreness` demands `ConnectedComponents`.

### CODE REVIEW

File: src/dachshund/simple_undirected_graph.rs [hunk 1]
```diff
@@ -14,13 +16,17 @@ use rand::distributions::WeightedIndex;
 use rand::prelude::*;
 use rand::seq::SliceRandom;
 use rand::Rng;
+<<<<<<< HEAD
```
Comment (mlmenard) [inline src/dachshund/simple_undirected_graph.rs:19]: Looks like there are some leftovers from a failed merge here.

**Review by mlmenard — CHANGES_REQUESTED** [review]
This is a big improvement!

2 things: Do you think it makes sense to have a separate algorithms folder (like use lib_dachshund::dachshund::algorithms::coreness)  instead of just having everything in dachshund?

Secondly, there are some leftovers from a failed merge. Can you fix those and update the PR so I can merge?

File: src/dachshund/simple_undirected_graph.rs (hunk 1 above)
Comment (bogdanstate) [reply src/dachshund/simple_undirected_graph.rs]: Resolved!

**Comment by bogdanstate** [comment]
@mlmenard, sorry for the long delay, replied to your comments and moved the code to an `algorithms` module.

**Review by mlmenard — APPROVED** [review]
Looks good to me now. Merging.

**Pull request merged**